### PR TITLE
build: separate requirements.txt for better layering

### PR DESCRIPTION
**Pull request by joshuarli**
The `COPY . /usr/src/snuba` layer is pretty much invalidated all the time, and the build step that follows fetches ~50 MB (227 MB unpacked) system packages. This will happen on like, every update commit to a PR which is pretty painful.

It's fixed by first installing requirements.txt before. To my knowledge at least uwsgi and the dogstatsd SO requires those packages to build.

The typical build time right now is a consistent 5m 30s, I'm not actually sure how much faster this is yet because `--cache-from getsentry/snuba:latest` and Cloud Build only pushes to the registry if it's a commit on master.  It probably won't be much faster, because the extra pip noop step is going to take a bit of time and Travis network is reasonably fast. But this is indeed way more respectful to Debian repositories.

Also includes some refactors from https://github.com/getsentry/getsentry/pull/4408, but most of those don't apply since this is a prod image.

### CODE REVIEW

**Comment by joshuarli** [comment]
The new [`us.gcr.io/sentryio/snuba:615a684d3bc4cac76693080b5ba6f939c11c5cc7`](https://console.cloud.google.com/gcr/images/sentryio/US/snuba@sha256:f0ac11db608139067950379f8d37ad85886a3f7108c8f3724265d3ea8889c3ba/details?tab=info) is `151.28 MB`, the old [`us.gcr.io/sentryio/snuba:edea3a458c01ce3b316e59b36d554dc91f9f8ca6 `](https://console.cloud.google.com/gcr/images/sentryio/US/snuba@sha256:030186894bd211451be5742520705b8361675bedee3afe2f5e999af1911087e3?tag=edea3a458c01ce3b316e59b36d554dc91f9f8ca6) is ... `107.67 MB`.

This is because docker pull will fetch all layers, not the final image. So that's the tradeoff here.

Even better would be to COPY the requirements file first, and have all those installed _before_ everything is copied. This is what we do for getsentry.

**Comment by joshuarli** [comment]
Okay, the image size is now 107.66 MB, pretty much the same.

https://console.cloud.google.com/gcr/images/sentryio/US/snuba@sha256:e6968af790bdd099b1d5351cf85fb9ca20d0fd5819132abe10eda78fa13341a5/details?tab=info

File: Dockerfile [hunk 1]
```diff
@@ -89,7 +93,5 @@ ENV SNUBA_RELEASE=$SNUBA_VERSION_SHA \
     UWSGI_DOGSTATSD_EXTRA_TAGS=service:snuba
 
 EXPOSE 1218
-
-COPY docker_entrypoint.sh ./
```
Comment (fpacifici) [inline Dockerfile:93]: Is this related to the change in the PR ?

File: Dockerfile [hunk 2]
```diff
@@ -74,9 +68,19 @@ RUN set -ex; \
...
     mv dogstatsd_plugin.so /var/lib/uwsgi/; \
     uwsgi --need-plugin=/var/lib/uwsgi/dogstatsd --help > /dev/null; \
-    snuba --help; \
     \
-    apt-get purge -y --auto-remove $buildDeps
+    apt-get purge -y --auto-remove $buildDeps; \
+    rm -rf /var/lib/apt/lists/*;
+
+# Layer cache is pretty much invalidated here all the time,
+# so try not to do anything heavy beyond here.
```
Comment (fpacifici) [inline Dockerfile:76]: This gets invalidated at every commit in snuba, right?

Comment (joshuarli) [reply Dockerfile:76]: Yes.

File: Dockerfile (hunk 1 above)
Comment (joshuarli) [reply Dockerfile:93]: It's not necessary, since everything is copied earlier.

**Review by fpacifici — APPROVED** [review]

**Comment by joshuarli** [comment]
Looks like builds sped up by about a minute. :)
